README.md
# [top-vhost](https://github.com/hex7c0/top-vhost)

[![NPM version](https://img.shields.io/npm/v/top-vhost.svg)](https://www.npmjs.com/package/top-vhost)
[![Linux Status](https://img.shields.io/travis/hex7c0/top-vhost.svg?label=linux)](https://travis-ci.org/hex7c0/top-vhost)
[![Windows Status](https://img.shields.io/appveyor/ci/hex7c0/top-vhost.svg?label=windows)](https://ci.appveyor.com/project/hex7c0/top-vhost)
[![Dependency Status](https://img.shields.io/david/hex7c0/top-vhost.svg)](https://david-dm.org/hex7c0/top-vhost)
[![Coveralls](https://img.shields.io/coveralls/hex7c0/top-vhost.svg)](https://coveralls.io/r/hex7c0/top-vhost)

Top-Down virtual host for [nodejs](http://nodejs.org/).
Optional, you can set this module like a proxy or a redirect.

## Installation

Install through NPM

```bash
npm install top-vhost
```
or
```bash
git clone git://github.com/hex7c0/top-vhost.git
```

## API

inside expressjs project
```js
var vhost = require('top-vhost');
var father = require('express')();
var child = require('express')();

child.get('/',function(req,res) {

  res.send('hello');
});

father.use(vhost({
  domain: 'foo.com',
  framework: child,
})).listen(3000);
```

### vhost(options)

#### options

 - `domain` - **String | RegExp** Name or Regex of virtual host *(required)*
 - `dynamic` - **String** Path of dynamic configuration json file (only with Proxies). Array of hosts *(default "disabled")*
 - `static` - **String** Path of static configuration json file. Preserve options *(default "disabled")*
 - `redirect` - **Array** Array of url that will be redirected to this domain *(default "disabled")*
 - `redirectStatus` - **Number** Set type of HTTP header for redirect (permanently 301 or temporary 307) or use another *(default "301")*
 - `insensitive` - **Boolean** Set True if want parse insensitive match *(default "sensitive")*
 - `stripWWW` - **Boolean** Stripe all "www." url *(default "disabled")*
 - `stripOnlyWWW` - **Boolean** Redirect all "www." url to "." url *(default "disabled")*
 - `stripHTTP` - **Boolean** Redirect all "http://" to "https://" url *(default "disabled")*
 - `stripHTTPS` - **Boolean** Redirect all "https://" to "http://" url *(default "disabled")*
 - `framework` - **Function** Functions related to this virtual host *(optional)*
 - `proxies` - **Object** Object for build http proxy, related to [`http-proxy`](https://github.com/nodejitsu/node-http-proxy) *(optional)*

related to https://github.com/nodejitsu/node-http-proxy/blob/master/lib/http-proxy.js#L29-L52

## Examples

Take a look at my [examples](examples)

### [License GPLv3](LICENSE)

History.md
v1.8.7 / 2016-02-13
==================

  * Tested against `node`@5.3
  * Update `http-proxy`@1.13.1

v1.8.6 / 2015-12-09
==================

  * Tested against`node`@5
  * Update `setheaders`@0.1.5

v1.8.5 / 2015-10-29
==================

  * Tested against`node`@4
  * Update `setheaders`@0.1.4
  * Update `http-proxy`@1.12.0

v1.8.4 / 2015-08-28
==================

  * Update `setheaders`@0.1.3

v1.8.3 / 2015-07-07
==================

  * Fix SPDX license
  * Improve `iojs` test
  * Update `http-proxy`@1.11.1

v1.8.1 / 2015-04-17
==================

  * Update `http-proxy`@1.10.1

v1.8.0 / 2015-03-23
==================

  * Update `http-proxy`@1.9.0
  * Update `setheaders`@0.1.1
  * `windows` test
  * `iojs` test
  * `coveralls` test

v1.7.19 / 2015-01-13
==================

  * Fix Header on `nodejs@0.11`

v1.7.18 / 2015-01-01
==================

  * Update devDependencies

v1.7.17 / 2014-12-21
==================

  * Update `http-proxy`@1.8.1

v1.7.16 / 2014-12-09
==================

  * Update `http-proxy`@1.7.3

v1.7.15 / 2014-12-04
==================

  * Update `http-proxy`@1.7.1

v1.7.14 / 2014-11-29
==================

  * Update `http-proxy`@1.7.0

v1.7.13 / 2014-11-13
==================

  * Update `http-proxy`@1.6.2

v1.7.12 / 2014-11-07
==================

  * Update `http-proxy`@1.6.1

v1.7.11 / 2014-11-02
==================

  * Update `http-proxy`@1.6.0
  * Test

v1.7.10 / 2014-10-27
==================

  * `jshint`

v1.7.9 / 2014-10-04
==================

  * Remove jsdoc
  * `package.json` min

v1.7.8 / 2014-10-03
==================

  * Update `http-proxy`@1.5.3

v1.7.7 / 2014-10-01
==================

  * Update `http-proxy`@1.5.2

v1.7.6 / 2014-09-30
==================

  * Update `http-proxy`@1.5.0

v1.7.5 / 2014-08-23
==================

  * Update `http-proxy`@1.4.3

v1.7.3 / 2014-08-23
==================

  * Performance tips

v1.7.1 / 2014-08-11
==================

  * Update README.md
  * Update `http-proxy`@1.2.0

v1.7.0 / 2014-08-04
==================

  * Wrapper function
  * Add stripOnlyWWW (options)

v1.6.0 / 2014-07-27
==================

  * Tested against"node": ">=0.10.0" only
  * Using task runner `grunt`
  * ".npmignore" more aggressive
  * `uglify` compiles

v1.5.3 / 2014-07-22
==================

  * Clean code

v1.5.2 / 2014-07-21
==================

  * Remove case [insensitive](http://jsperf.com/case-sensitive-regex-vs-case-insensitive-regex)
  * Add `insensitive` option
  * Redirect builder with custom status code

v1.5.1 / 2014-07-21
==================

  * Remove unused var

v1.5.0 / 2014-07-09
==================

  * Improve loop performance
  * try-catch responsibility of the user

v1.4.0 / 2014-07-06
==================

  * Add `redirectStatus` option
  * Add `stripWWW` option
  * Add `stripHTTP` option
  * Add `stripHTTPS` option
  * Small fixies

v1.3.2 / 2014-07-04
==================

  * Return Boolean

v1.3.1 / 2014-07-03
==================

  * `file` option is deprecated

v1.3.0 / 2014-07-03
==================

  * Rename `file` option to `dynamic`
  * Add `static` option
  * Redirect even url
  * Subdivision in functions

v1.2.0 / 2014-07-01
==================

  * Add `file` option

v1.1.1 / 2014-06-30
==================

  * Better performance

v1.1.0 / 2014-06-29
==================

  * Add redirect (option)

v1.0.0 / 2014-06-29
==================

  * HTPP proxy

v0.0.1 / 2014-06-29
==================

  * Project start


index.js
'use strict';
/**
 * @file top-vhost main
 * @module top-vhost
 * @subpackage main
 * @version 1.8.0
 * @author hex7c0 <hex7c0@gmail.com>
 * @copyright hex7c0 2014
 * @license GPLv3
 */

/*
 * initialize module
 */
var setHeader = require('setheaders').setProctedHeader;

/*
 * functions
 */
/**
 * main
 * 
 * @function vhost
 * @export vhost
 * @param {Object} opt - various options. Check README.md
 * @return {Function}
 */
function vhost(opt) {

  /*
   * wrapper module
   */
  var redirect = redirect_body(301);
  var insensitive;

  /*
   * common functions
   */
  /**
   * http redirect builder
   * 
   * @function builder
   * @param {Array} moved - url
   * @param {String} domain - req.headers.host
   * @return {Object}
   */
  function builder(moved, domain) {

    for (var i = 0, ii = moved.length; i < ii; ++i) {
      moved[i] = expression(moved[i]);
    }
    return {
      reg: moved,
      orig: domain,
    };
  }

  /**
   * http redirect builder
   * 
   * @function redirect_body
   * @param {Integer} res - response code
   * @return {Function}
   */
  function redirect_body(code) {

    var cod = code;

    /**
     * http redirect status code
     * 
     * @function
     * @param {Object} res - response
     * @param {Object} moved - regex url
     * @param {String} host - req.headers.host
     * @param {String} url - req.url
     */
    return function(res, moved, host, url) {

      var reg = moved.reg;

      for (var i = 0, ii = reg.length; i < ii; ++i) {
        if (reg[i].test(host)) {
          var to = moved.orig + url;
          setHeader(res, 'Location', to); // 0.11
          res.writeHead(cod, { // 0.10
            Location: to,
          });
          res.end();
          return true;
        }
      }
      return false;
    };
  }

  /**
   * regex builder
   * 
   * @function expression
   * @param {String} urls - url to be parsed
   * @return {RegExp}
   */
  function expression(urls) {

    var url = urls.replace(/http[s]?:\/\//i, '').replace(/\*/g, '([^.]+)');
    // add starting index
    if (url[0] != '^') {
      url = '^' + url;
    }
    // if (url[url.length - 1] === '$') {
    // url = url.slice(0,url.length - 1);
    // }
    return new RegExp(url, insensitive);
  }

  /*
   * return functions
   */
  /**
   * strip snippet
   * 
   * @function strip
   * @param {Array} moved - http(s) redirect
   * @return {Function}
   */
  function strip(moved) {

    return function vhost(req, res, next) {

      if (!redirect(res, moved, req.headers.host, req.url)) {
        next();
      }
      return;
    };
  }

  /**
   * proxy work
   * 
   * @function proxies
   * @param {RegExp} domain - vhost
   * @param {Array} moved - array of redirect
   * @param {Object} proxy - proxy
   * @return {Function}
   */
  function proxies(domain, moved, proxy) {

    var domainC = domain, proxyC = proxy;
    if (moved) {
      var mvd = moved, rdc = redirect;
      return function vhost(req, res, next) {

        var host = req.headers.host;
        if (domainC.test(host)) {
          proxyC.web(req, res);
          return;
        } else if (rdc(res, mvd, host, req.url)) {
          return;
        }
        return next();
      };
    }
    return function vhost(req, res, next) {

      var host = req.headers.host;
      if (domainC.test(host)) {
        return proxyC.web(req, res);
      }
      return next();
    };
  }

  /**
   * framework work
   * 
   * @function framework
   * @param {RegExp} domain - vhost
   * @param {Array} moved - array of redirect
   * @param {Object} fw - framework
   * @return {Function}
   */
  function framework(domain, moved, fw) {

    var domainC = domain, fwC = fw;
    if (moved) {
      var mvd = moved, rdc = redirect;

      return function vhost(req, res, next) {

        var host = req.headers.host;
        if (domainC.test(host)) {
          return fwC(req, res);
        } else if (rdc(res, mvd, host, req.url)) {
          return;
        }
        return next();
      };
    }
    return function vhost(req, res, next) {

      var host = req.headers.host;
      if (domainC.test(host)) {
        return fwC(req, res);
      }
      return next();
    };
  }

  /**
   * dynamic file
   * 
   * @function dynamic
   * @param {String} files - input file
   * @return {Function}
   */
  function dynamics(files) {

    var rdc = redirect, exp = expression;
    var file = require('path').resolve(files);
    var fs = require('fs');
    var http_proxy = require('http-proxy');
    if (!fs.existsSync(file)) {
      throw new Error(file + ' not exists');
    }

    return function vhost(req, res, next) {

      var host = req.headers.host;
      // file refresh; instead require(), that use cache
      var data = JSON.parse(fs.readFileSync(file, 'utf8'));
      for (var i = 0, ii = data.length; i < ii; ++i) {
        var moved;
        var d = data[i];
        if (Boolean(d.insensitive)) {
          insensitive = 'i';
        } else {
          insensitive = undefined;
        }
        var domain = exp(d.domain.source || d.domain);
        var proxy = http_proxy.createProxyServer(d.proxies);
        if (d.redirect) {
          moved = builder(d.redirect, d.domain.source || d.domain);
          if (domain.test(host)) {
            proxy.web(req, res);
            return true;
          }
          if (rdc(res, moved, host, req.url)) {
            return true;
          }
        }
        if (domain.test(host)) {
          proxy.web(req, res);
          return true;
        }
      }
      return next();
    };
  }

  /**
   * statics file
   * 
   * @function statics
   * @param {String} files - input file
   * @param {Object} obj - parsed options
   * @return {Function}
   */
  function statics(files, obj) {

    var file = require('path').resolve(files);
    if (!require('fs').existsSync(file)) {
      throw new Error(file + ' not exists');
    }
    var d = require(file);
    var domain, proxy;
    var moved, temp;

    // domain
    if (Boolean(d.insensitive)) {
      insensitive = 'i';
    } else {
      insensitive = undefined;
    }
    if (obj.domain) {
      domain = obj.domain;
    } else if ((domain = d.domain)) {
      if (domain.constructor.name === 'RegExp') {
        domain = domain.source;
      } else if (typeof (domain) != 'string') {
        throw new Error('invalid "domain" argument');
      }
      domain = expression(domain);
    } else if (d.dynamic) {
      // pass
    } else {
      throw new Error('"domain" is required');
    }

    // single
    if (obj.moved) {
      moved = obj.moved;
    } else if (Array.isArray(d.redirect) === true) {
      moved = builder(d.redirect, obj.orig || d.domain.source || d.domain);
    }
    if ((temp = Number(d.redirectStatus))) {
      redirect = redirect_body(temp);
    }

    // extra
    if (d.stripWWW || d.stripOnlyWWW || d.stripHTTP || d.stripHTTPS) {
      temp = obj.orig || d.domain.source || d.domain;
      if (d.stripHTTP) {
        return strip({
          reg: [ /./ ],
          orig: temp.replace(/http:\/\//i, 'https://'),
        });
      }
      if (d.stripHTTPS) {
        return strip({
          reg: [ /./ ],
          orig: temp.replace(/https:\/\//i, 'http://'),
        });
      }
      if (d.stripOnlyWWW) {
        return strip({
          reg: [ /^www./i ],
          orig: temp.replace(/www./i, '.'),
        });
      }
      if (d.stripWWW) {
        if (typeof (moved) != 'object') {
          moved = {
            reg: [ /^www./i ],
            orig: domain,
          };
        } else {
          moved.reg.push(/^www./i);
        }
      }
    }

    // return
    if (d.dynamic) {
      return dynamics(String(d.dynamic));
    } else if (obj.framework) {
      return framework(domain, moved, obj.framework);
    } else if (obj.proxies) {
      return proxies(domain, moved, obj.proxies);
    } else if (d.proxies && typeof (d.proxies) === 'object') {
      proxy = require('http-proxy').createProxyServer(d.proxies);
      return proxies(domain, moved, proxy);
    }
    throw new Error('"framework" or "proxies" are required');
  }

  var options = opt || Object.create(null);
  var domain, fw, proxy;
  var moved, temp;
  var next = Object.create(null);

  // exclusive
  if (options.framework && typeof (options.framework) === 'function') {
    fw = options.framework;
    next.framework = fw;
  } else if (options.proxies && typeof (options.proxies) === 'object') {
    proxy = require('http-proxy').createProxyServer(options.proxies);
    next.proxies = proxy;
  }

  // domain
  if (Boolean(options.insensitive)) {
    insensitive = 'i';
  } else {
    insensitive = undefined;
  }
  domain = options.domain;
  if (domain) {
    if (domain.constructor.name === 'RegExp') {
      domain = domain.source;
    } else if (typeof (domain) != 'string') {
      throw new Error('invalid "domain" argument');
    }
    domain = expression(domain);
    next.domain = domain;
    next.orig = options.domain.source || options.domain;
  } else if (options.dynamic || options.static) {
    // pass
  } else {
    throw new Error('"domain" is required');
  }

  // single
  if (options.file) {
    console.info('top-vhost > "file" option is deprecated');
  }
  if (Array.isArray(options.redirect) === true) {
    moved = builder(options.redirect, options.domain.source || options.domain);
    next.moved = moved;
  }
  if ((temp = Number(options.redirectStatus))) {
    redirect = redirect_body(temp);
  }

  // extra
  if (options.stripWWW || options.stripOnlyWWW || options.stripHTTP
    || options.stripHTTPS) {
    temp = options.domain.source || options.domain;
    if (options.stripHTTP) {
      return strip({
        reg: [ /./ ],
        orig: temp.replace(/http:\/\//i, 'https://'),
      });
    } else if (options.stripHTTPS) {
      return strip({
        reg: [ /./ ],
        orig: temp.replace(/https:\/\//i, 'http://'),
      });
    } else if (options.stripOnlyWWW) {
      return strip({
        reg: [ /^www./i ],
        orig: temp.replace(/www./i, '.'),
      });
    }
    if (options.stripWWW) {
      if (typeof (moved) != 'object') {
        moved = {
          reg: [ /^www./i ],
          orig: domain,
        };
      } else {
        moved.reg.push(/^www./i);
      }
    }
  }

  // return
  if (options.dynamic) {
    return dynamics(String(options.dynamic));
  } else if (options.static) {
    return statics(String(options.static), next);
  } else if (fw) {
    return framework(domain, moved, fw);
  } else if (proxy) {
    return proxies(domain, moved, proxy);
  }
  throw new Error('"framework" or "proxies" are required');
}
module.exports = vhost;
